src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,545 @@
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $ignoreCase = false;
+
+    abstract public function __construct(string $string = '');
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function after($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOf($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if (!$includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice($i);
+    }
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function afterLast($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOfLast($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if (!$includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice($i);
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function append(string ...$suffix): self;
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function before($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOf($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function beforeLast($needle, bool $includeNeedle = false): self
+    {
+        if (null === $i = $this->indexOfLast($needle)) {
+            $str = clone $this;
+            $str->string = '';
+
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $str = clone $this;
+            $str->string = $needle;
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @return static
+     */
+    public function camel(): self
+    {
+        $str = clone $this;
+        $str->string = lcfirst(str_replace(' ', '', ucwords(preg_replace('/[^a-zA-Z0-9\x7f-\xff]++/', ' ', $this->string))));
+
+        return $str;
+    }
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @param string|string[] $needle
+     */
+    public function contains($needle): bool
```

Comment:
i believe this is the place for either a variadic argument ( `string ...$needles` ) or an another `$offset` argument, to allow either : `$str->contains('a', 'b');` or  `$str->contains('a', $offset);`.

the preferred API IMO would be : 

```
contains(string $needle, int $offset = 0): bool;
containsAny(array<string> $needles, int $offset = 0): bool;
containsAll(array<string> $needles, int $offset = 0): bool;
```

when doing `$str->contains(['a', 'b'])` is not clear whether it is checking if `$str` contains 1 or all of the needles without looking at the implementation.

Reply:
I'm not up for using variadics: they close extensibility, require unusual syntax, don't work with associative arrays.

> when doing $str->contains(['a', 'b']) is not clear whether it is checking if $str contains 1 or all of the needles

Without data to back such statements, we cannot conclude anything. [Others do](https://github.com/laravel/framework/blob/5.5/src/Illuminate/Support/Str.php#L100) and all if fine.

I'm not for adding more methods.

Reply:
@nicolas-grekas I've seen you dismiss any variadic function suggestions on APIs for quite literally years now, so I'm not expecting you to change your mind (though it would be great if such ideas weren't simply dismissed and actually discussed or voted on by the community at large), but I take issue with both of the arguments you continue to use against variadic functions.

1. **They close extensibility.** This only matters when future extensibility makes sense. A function like `contains`, where the sole purpose is to check if "the string contains any of the arguments passed in", doesn't have a need for future extension of functionality. If a feature request was later submitted to enable the ability to perform the same check, but only after a given offset, it could easily be added as its own method (which, IMHO, can often a better decision, resulting in a cleaner and simpler API, than continually adding optional arguments to perform tasks outside the original function scope).

2. **Require unusual syntax.** First, this is entirely subjective, and moreover, it's not a commonly held opinion (countless languages wouldn't themselves use a strikingly similar syntax if many others shared this belief). That being said, I do not even understand this position.

```php
// using variadic functions is simple...
variadicFunc(1, 2, 3, 4);
variadicFunc(...[1, 2, 3, 4]);
$numbers = [1, 2, 3, 4];
variadicFunc(...$numbers);
variadicFunc(...array_map(function (int $i) {
    return $i + 1;
}, $numbers));

// and not more complex than passing to an array argument
arrayFunc([1, 2, 3, 4]);
arrayFunc([1, 2, 3, 4]);
$numbers = [1, 2, 3, 4];
variadicFunc($numbers);
variadicFunc(array_map(function (int $i) {
    return $i + 1;
}, $numbers));
```

2. **Doesn't work with associative arrays.** Since you are a highly experienced PHP programmer, I'm sure it will come as to no surprise to you that this statement is obviously misleading. While associative array cannot be unpacked using `...` on the array directly, in the unusual (or, at the very least, uncommon) situation where an associative array is being passed, the solution is quite simple and straightforward.

```php
$associative = ['one' => 1, 'two' => 2, 'three' => 3, 'four' => 4];
variadicFunc(...array_values($numbers));
```

It's *absolutely fine* if you just *don't like variadic functions*, but Symfony is made for the community and not any one single person. __I'm not trying to imply variadic function are great for every API and in every situation__ but they seem to be universally ignored in Symfony, and I don't believe that choice is in the best interest of its users.

Reply:
> This only matters when future extensibility makes sense

By definition, we don't know the future. Any assumptions one can make today is going to be wrong the next days. Actually, yours is already wrong: the `contains()` method now has a second argument, which helps solve some problems in a performant way (skipping memcopies).

> variadicFunc(...array_values($numbers));

that's two array copies instead of zero when variadics are not used.

Reply:
@nicolas-grekas Just to provide some additional context, I didn't intend for my post to come off quite as "pointed" as it did on re-reading it; I love the component itself (regardless of API choices) and respect everything you do for Symfony. I am simply annoyed at the continued dismissive attitude toward variadic functions, but please don't take anything the wrong way. :-)

Reply:
Thank you for your kind words.

I feel like some are hacking variadics to work around the fact the PHP type system doesn't handle collections yet. When considering them, the question that resonates to me is: would I still use variadics if PHP were to support type collections and polymorphic methods? If the answer is no, then using them would be abusing the API. e.g. `sprintf()` is a perfect use for variadics. But here, it's not to me (the 2nd $offset argument kinda proves the point now :) )

Reply:
> when doing $str->contains(['a', 'b']) is not clear whether it is checking if $str contains 1 or all of the needles without looking at the implementation

Fully agree.

Reply:
OK, I'm removing the method: it creates more problems than it solves. Here is the replacement:
`null !== $this->indexOf($needle, $offset)`